src/front/glsl/parser/declarations.rs

```diff
@@ -198,7 +198,9 @@ impl<'source> ParsingContext<'source> {
                 let is_const = ctx.qualifiers.storage.0 == StorageQualifier::Const;
 
                 match parser.solve_constant(ctx.ctx, root, meta) {
-                    Ok(res) => Some(res),
+                    // Note that even if we solve the constant, we only use the solution for global constants.
+                    // Locals have to be re-assigned every time they are encountered.
+                    Ok(res) if ctx.external => Some(res),
```

Comment:
This check should be moved before the constant is solved, no point in wasting time processing it if it isn't going to be used